lib/src/revset.rs

```diff
@@ -1622,13 +1623,51 @@ impl<'index> Iterator for ReverseRevsetIterator<'index> {
     }
 }
 
+struct RevsetIteratorState<'index> {
+    iter: Box<dyn Iterator<Item = IndexEntry<'index>> + 'index>,
+    entries: Vec<IndexEntry<'index>>,
+}
+
```

Comment:
Maybe there would be some reason behind this design decision, but I think `contains()`, `len()`, etc. can be implemented as an iterator adapter like peekable. It will eliminate self referencing struct and mutex, and the caller will pay the cost only when needed.

If Google's internal backend provides an efficient `contains()` implementation, the iterator adapter can be abstracted by a trait object, I think.

```
Revset {
    fn iter() -> ... // trivial iterator

    fn cachable_iter(&self, index) -> Box<dyn ... + '_> 
    // maybe the returned object is close to the new Revset trait introduced in this PR?
}
```

Reply:
We just had an internal meeting (@torquestomp, @spectral54, @jaraco, and I) about what we really need from the revset API. The most important goal is to be able to be able to reimplement [RevsetGraphIterator](https://github.com/martinvonz/jj/blob/main/lib/src/revset_graph_iterator.rs) to not walk commits outside of the set one step at a time outside as it currently does.

I had planned to do it like [Sapling does it](https://github.com/facebook/sapling/blob/d09f4cf56afd4a5a8c94449cffe3a48a010b1424/eden/scm/edenscm/graphmod.py#L79-L95). However, the conclusion from the meeting was that it would be simpler and maybe more efficient to instead move the graph iterator, too, into the index. `Revset` currently has an `iter()` method, which has a `.graph()` adapter, but if we instead add a `iter_graph()` method to `Revset`, that will give us more flexibility to implement it as we want on our server.

Doing it this way is less flexible, but it seems that we don't need the flexibility for anything else right now anyway. We can revisit if other usecases than the graph iterator come up.

Note that the graph iterator currently has an option to include or skip transitive edges, and the new `iter_graph()` would just always skip transitive edges (since that's what jj currently does). Maybe our custom implementation will or maybe it won't. If users (at Google or outside) later decide that they want the option to include transitive edges, we can add an argument to the function.

The default implementation of `iter_graph()` would use the current graph iterator for now, but maybe we'll make it work more like Sapling's implementation later. It performs well enough on repos up to the size of the Linux repo anyway.

What do you think?

Reply:
> it would be simpler and maybe more efficient to instead move the graph iterator, too, into the index. `Revset` currently has an `iter()` method, which has a `.graph()` adapter, but if we instead add a `iter_graph()` method to `Revset`, that will give us more flexibility to implement it as we want on our server.
> 
> Doing it this way is less flexible, but it seems that we don't need the flexibility for anything else right now anyway. We can revisit if other usecases than the graph iterator come up.

That sounds good to me.

Just to clarify, I don't think it means we'll need a shared cache across multiple revset iterators. If we rewrite the default RevsetGraphIterator to be more Sapling-like, it would just need an iterator + set op, which I think can be implemented with an exclusively-owned RevsetIteratorState.

Reply:
Yes, I think that's correct. If the RevsetGraphIterator knows the type of iterator it's backed by, it can use that (the RevsetIteratorState, as you said) without needing generic set operations on revsets.